File: src/GCov/Data/CoverageFunctionData.java
package GCov.Data;

import java.util.Map;
import java.util.TreeMap;

public class CoverageFunctionData {

    private final Map<Integer, CoverageLineData> m_lines = new TreeMap<>();
    private final int m_startLine;
    private final int m_endLine;
    private int m_executionCount;
    private final String m_functionName;
    private CoverageFileData m_fileData;

    public CoverageFileData getFileData() {
        return m_fileData;
    }

    void setFileData(CoverageFileData fileData) {
        m_fileData = fileData;
    }

    CoverageFunctionData(int startLine, int endLine, int executionCount, String functionName) {
        m_startLine = startLine;
        m_endLine = endLine;
        m_executionCount = executionCount;
        m_functionName = functionName;
    }

    void emplaceLine(int lineNumber, int executionCount, boolean unexecutedBlock) {
        this.emplaceLine(new CoverageLineData(lineNumber, executionCount, unexecutedBlock));
    }

    private void emplaceLine(CoverageLineData lineData) {
        CoverageLineData line = m_lines.get(lineData.getLineNumber());
        if (line == null) {
            m_lines.put(lineData.getLineNumber(), lineData);
            m_fileData.getLineData().put(lineData.getLineNumber(),lineData);
        } else {
            line.setExecutionCount(line.getExecutionCount() + lineData.getExecutionCount());
        }
    }

    boolean lineIsInFunction(int line) {
        return line >= m_startLine && line <= m_endLine;
    }

    public int getStartLine() {
        return m_startLine;
    }

    int getExecutionCount() {
        return m_executionCount;
    }

    void setExecutionCount(int executionCount) {
        m_executionCount = executionCount;
    }

    String getFunctionName() {
        return m_functionName;
    }

    public int getCoverage() {
        int coverage = 0;
        for(Map.Entry<Integer, CoverageLineData> entry : m_lines.entrySet()) {
            if(entry.getValue().getExecutionCount() > 0) {
                coverage++;
            }
        }
        return coverage;
    }

    public Map<Integer, CoverageLineData> getLines() {
        return m_lines;
    }

    @Override
    public String toString() {
        return m_functionName;
    }
}

File: src/gcov/data/CoverageFunctionData.kt
package gcov.data

import java.util.TreeMap

class CoverageFunctionData(var startLine: Int = -1, var endLine: Int = -1, var functionName: String = "<unknown function>", var executionCount: Int = 0) {

    private val myLines = TreeMap<Int, CoverageLineData>()

    var fileData: CoverageFileData? = null
        internal set

    constructor(fileData: CoverageFileData) : this(){
        this.fileData = fileData
    }

    val coverage: Int
        get() = myLines.values.count{it.executionCount > 0}

    val lines: Map<Int, CoverageLineData>
        get() = myLines

    fun emplaceLine(lineNumber: Int, executionCount: Int, unexecutedBlock: Boolean) {
        this.emplaceLine(CoverageLineData(lineNumber, executionCount, unexecutedBlock))
    }

    fun emplaceLine(lineData: CoverageLineData) {
        val line = myLines[lineData.lineNumber]
        if (line == null) {
            myLines[lineData.lineNumber] = lineData
            fileData?.lineData?.put(lineData.lineNumber, lineData)
        } else {
            line.executionCount += lineData.executionCount
        }
    }

    fun lineIsInFunction(line: Int): Boolean {
        return line in startLine..endLine
    }

    override fun toString(): String {
        return functionName
    }
}